src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -372,7 +372,106 @@ protected void InitializeReferences()
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        private static Match? InternalPerformScan(bool quick, string input, int beginning, RegexRunner runner, ReadOnlySpan<char> span, bool returnNullIfQuick)
+        {
+            // We need to save the original beginning of the input in case we are in the CompileToAssembly case, which relies on
+            // setting runtextbeg and runtextend correctly. We cannot calculate these values on the fly based on input and the span
+            // because it is not guaranteed that span.Length + beginning = input.
+            runner.originalRuntextbeg = beginning;
+            runner.Scan(span);
+
+            Match? match = runner.runmatch;
+            // if we got a match, set runmatch to null if quick is true
+            if (match!.FoundMatch)
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+                if (match.Text != input)
+                {
+                    match.Text = input;
+                }
+
+                if (quick && returnNullIfQuick)
+                {
+                    runner.runmatch!.Text = null; // Drop reference to text
+                    return null;
+                }
+
+                if (!quick)
+                    runner.runmatch = null;
+
+                match.Tidy(runner.runtextpos);
+
+                // If there was a match and the original text was sliced, then add beginning to the index to get the real
+                // Index of the match.
+                if (match.Success && beginning != 0)
+                {
+                    match.AddBeginningToIndex(beginning);
+                }
+
+                return match;
+            }
+
+            runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+            if (!quick)
+            {
+                runner.runmatch = null;
+            }
+            else
+            {
+
+                if (runner.runmatch != null)
+                {
+                    runner.runmatch.Text = null;
+                }
+            }
+
+            return RegularExpressions.Match.Empty;
+        }
+
+        internal Match? Run(ReadOnlySpan<char> input, int startat)
+        {
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
+            }
+
+            RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
+            try
+            {
+                runner.InitializeTimeout(internalMatchTimeout);
+                runner.InitializeForScan(this, input, startat, quick: true);
+
+                runner.Scan(input);
+
+                return runner.runmatch!.FoundMatch ? null : RegularExpressions.Match.Empty;
```

Comment:
Technically we probably can't assume runmatch is non-null here, since a derived runner could have set it to null.  I don't think I care, though, since there's a lot of things a RegexRunner could do that will result in null refs. Just calling it out.

Reply:
I can still adjust the condition and also return null when runmatch is null.